All my sequencer refactorings, with commentary

adroitwhiz opened the pull request:
### Resolves

Resolves https://github.com/LLK/scratch-vm/issues/3031
Resolves https://github.com/LLK/scratch-vm/issues/3032
Resolves https://github.com/LLK/scratch-vm/issues/3033
Resolves https://github.com/LLK/scratch-vm/issues/3034

Also removes a lot of tech debt and improves performance.

### Proposed Changes

This PR contains all the refactoring work I've done on the block sequencing and execution code, including the work I did to re-land mzgoddard's performance PRs #2145 and #2148. All commits, including his, contain my (hopefully) detailed commentary on what each change does.

### Reason for Changes

The VM is currently a massive pile of tech debt. Many of the people who originally worked on huge portions of the sequencer are no longer at Scratch, and the current code is hard to understand. My hope is that these refactors make it clearer how the sequencer works, both to anyone on the current team and to any third parties who want to replicate its behavior when implementing their own versions of Scratch.

These changes will also make it much easier to add future functionality to the sequencer--[event blocks with droppable inputs](https://github.com/LLK/scratch-vm/pull/2945), custom reporter blocks, etc. are all very hard to implement given the current state of the codebase.

Reviewing these changes would also be a good way to learn about the sequencer and how it works. I've done my best to make each change atomic, and I explain the previous spaghetti code that I'm removing at each step, so it might explain the sequencer better than just reading the current code.

### Test Coverage

This PR includes tests that mzgoddard added in #2145, as well as my own tests to ensure that https://github.com/LLK/scratch-gui/issues/5011 doesn't reoccur.

mzgoddard committed:
refactor: remove empty procedures_definition block

With a few minor tweaks, the engine completely supports blocks with no
functions. This means we don't have to define no-op functions for blocks
that do nothing, like procedures_definition.

(original commit by mzgoddard, message reworded by adroitwhiz)

mzgoddard committed:
refactor: elevate top stack frame into thread

We access the top of the stack a lot. To make this faster, convert the
top stack block ID and stack frame to properties. The top stack block ID
is now Thread.pointer and the top stack frame is now Thread.stackFrame.
This requires some refactoring work--all places that checked whether
the stack length > 0 now need to check if the top stack frame or block
is null.

We also now copy procedure params further "down" stack frames so that
argument_reporter_* blocks only need to look at the top stack frame.
When we push a stack frame, we set its `params` value to that of its
parent. This should make accessing procedure parameters much faster.

Stack frames also have a new needsReset property. Since we don't often
do things that require the stack frame to be reset, the extra cost of
setting needsReset = true is outweighed by not having to reset the
stack frame every time we go to the next block.

Finally, procedures_call no longer sets or checks stackFrame.executed,
which seems to be a no-longer-needed workaround for procedure blocks
re-executing themselves dating back to 2016. It also now calls
startProcedure *before* initializing parameters instead of after. This
previously caused procedures to break because the promise-handling
code popped the stack too many times, but that code was removed in a
previous commit, and a regression test will be added in a later commit.

(original commit by mzgoddard, message reworded by adroitwhiz)

mzgoddard committed:
refactor: remove Thread.isKilled, rework stepThreads

We don't need to filter out stopped threads anymore, since we do so
in stepThreads ever since PR #1683. This means that Thread.isKilled is
no longer accessed from anywhere and can be removed.

We can also rearrange the order of the branches in stepThreads for
clarity and to put more often-taken branches first. The status === DONE
check in particular doesn't also need to check if the stack is length 0,
since we always set the status to DONE as soon as we reach the end of
the stack.

Finally, instead of tracking the number of active threads in
stepThreads, we only need to check if at least one thread is active.

(original commit by mzgoddard, message reworded by adroitwhiz)

mzgoddard committed:
refactor: add and use unwrapStack and stepHat

The unwrapStack function factors out the sequencer's functionality of
repeatedly popping the stack until we reach a block or the top of the
stack, and handling control flow that may happen when doing so.

stepThread has been significantly reworked to make use of this. If the
top of the stack is advanced to a null block, either by a hat block
being executed before entering stepThreads or by a promise resolving,
we use this new unwrapStack function to properly handle control flow
instead of just popping the stack once. Since the sequencer handles this
case, we don't have to unwrap the stack in `execute` after resolving a
promise since the sequencer will do it for us.

Additionally, we add a new stepHat function which executes one block on
a thread then advances to the next block. This allows the runtime to
call stepHat on the thread, meaning it doesn't have to import `execute`.

Finally, a small typo in retireThread has been corrected: it now
properly sets thread.stackFrames to an empty array (previously it was
setting "thread.stackFrame" to an empty array).

(original commit by mzgoddard, message reworded by adroitwhiz)

mzgoddard committed:
refactor: set blockGlowInFrame in execute.js

We were previously setting requestScriptGlowInFrame in execute.js, but
setting blockGlowInFrame in sequencer.stepThread. We now set them both
in execute.js for consistency.

(original commit by mzgoddard, message reworded by adroitwhiz)

mzgoddard committed:
test: unit test more stepToBranch scenarios

mzgoddard committed:
test: test stopThisScript in procedures

Test that stopThisScript in a procedure stays in the procedure and
moves it to the end where Sequencer can pop it normally

mzgoddard committed:
fix: fix stopThisScript inside procedures

It appears this might have been what stackFrame.executed was for.
If we don't push the null onto the stack when stopping a script, the
sequencer will see that we've done control flow and not move to the next
block. Because the top of the stack wasn't null, however, we never ended
up unwrapping the stack, and got stuck executing the same procedure
block. By pushing a null to the stack, we cause the sequencer to unwrap
the stack and go to the next block. This is hacky and will be changed
in a much later commit.

(original commit by mzgoddard, message reworded by adroitwhiz)

adroitwhiz committed:
refactor: remove dead code path from getCached

There was a code path that handled the case where getCached was not
passed the `CacheType` parameter. However, all call sites *do* pass that
parameter, and the JSDoc indicates that the parameter is required. We
can thus remove that code path.

adroitwhiz committed:
refactor: don't cache hat block inputs

The RuntimeScriptCache is used by Runtime.startHats to filter only the
hat blocks matching the specified opcode and *fields*, aka non-droppable
menus on blocks
(e.g. 'event_whenkeypressed', where {'KEY': 'SPACE'}).
Each RuntimeScriptCache object holds data related to one hat block: its
opcode and uppercase versions of all its fields (so that field matching
is case-insensitive: you can broadcast 'MeSsAgE1' and 'when I receive
message1' will still fire).

Previously, this cache also held droppable menus, aka *inputs*. These
are actually blocks in their own right, known as "shadow blocks". They
are essentially reporter blocks that store their reported value in a
specific field. The RuntimeScriptCache was copying this value out of the
field and into the `fieldsOfInputs` property.

However, this broke if you dropped a different reporter block into that
input--you'd have no way of knowing what it would evaluate to. Thus, all
hat blocks which use droppable inputs are "edge-activated"--they are
evaluated once per frame by the runtime as opposed to being activated by
the `startHats` function. Thus, storing inputs is not only unnecessary
but would never have worked anyway.

adroitwhiz committed:
refactor: rename fieldsOfInputs to fields

We're not storing fields of inputs anymore, only fields of the hat block
itself. We should thus rename the variable to what it actually does now.

adroitwhiz committed:
refactor: clean up field-cloning logic

Remove all the multiple-assignment and "check if this value has a
function that only strings have" stuff.

adroitwhiz committed:
refactor: remove BlockCached._fields

We wrote to this field in the BlockCached constructor, but never
actually access it anywhere.

adroitwhiz committed:
refactor: rename getInputs to getNonBranchInputs

This more accurately reflects what the function *actually* does. It
returns all inputs that aren't branches.

adroitwhiz committed:
refactor: stop caching block inputs in Blocks

The BlockCached class caches inputs, so we don't need to do so in the
Blocks class as well--the caches are all cleared at the same time, so
every time a BlockCached asked for block inputs, it would've been a
cache miss anyway.

adroitwhiz committed:
refactor: pass block directly into BlockCached

This removes an object allocation and simplifies things a bit.
BlockCached is now reaching into the block, but there are loads of other
functions that do so as well, and this makes things cleaner.

adroitwhiz committed:
refactor: set BlockCached._isShadowBlock via block

No need to guess when a block is a shadow block--it will tell us.

adroitwhiz committed:
refactor: remove BlockCached.fields and mutation

These were also only used inside the constructor function, and can be
changed into local variables as well.

adroitwhiz committed:
refactor: remove BlockCached._definedBlockFunction

This was only used from inside the constructor function, and hence can
be changed into a local variable instead.

adroitwhiz committed:
refactor: remove BlockCached._inputs and inputs

As with the previous commits, these don't need to be member variables.

adroitwhiz committed:
refactor: remove Thread.waitingReporter

I believe we used to use this to handle promise blocks, but the current
code never sets it to anything other than null or uses it in any way/

adroitwhiz committed:
test: test that params work with promise blocks

This test ensures that scratch-gui#5011 will not reoccur.

adroitwhiz committed:
test: add "escape from branch" promise test

adroitwhiz committed:
fix: don't use retireThread in unwrapStack

retireThread sets requestScriptGlowInFrame to false, which results in
stacks that only execute for one frame never glowing. The blocks would
execute, the end of the stack would be reached within one frame, and
the stack would be unwrapped, causing the script to never glow.
retireThread also empties stack and stackFrames, which is redundant
because we were only calling it if the stack was already empty.

adroitwhiz committed:
refactor: don't pop stack on rejected promises

This was kind of like stopping the script but not really, and likely
would break something if it actually happened. It's a lot easier to
reason about by just returning an empty string and then doing the same
thing we do with a successfully executed promise. If a block throws an
error, that's a bug in the block itself, and we don't care what we do
as long as we don't blow up.

adroitwhiz committed:
refactor: clarify promise thaw iteration a bit

It took me way too long to think through all the cases in which `i`
changed or didn't change. This is much clearer.

adroitwhiz committed:
refactor: rename opCached to oldOpID

This makes it clearer that we're talking about a block ID and not a
function.

adroitwhiz committed:
refactor: push reported value only when handling promises

We only read thread.justReported when resolving a promise, so we only
need to set it if we're waiting for a promise to resolve.

adroitwhiz committed:
refactor: disentangle handleReport from promises

Previously, handleReport did two different things at once, and the side
effects of one would just be ignored depending on the code path.

If it was being called from handlePromise, it would push the reported
value to the thread and set its status to RUNNING.
If it was being called from execute, it would check hat predicates,
update a monitor, and possibly display a reporter balloon.

Now, it only does the latter-the former has been moved to handlePromise.

adroitwhiz committed:
refactor: api around setting "thawed" promise values

They're now stored on threads instead of the stack frame, and I want to
be able to tell if a promise just resolved without it having to report a
value, so split it into "did resolve" and "resolved value", and change
the functions accordingly.

adroitwhiz committed:
refactor: don't handle control flow in handlePromise

Previously, handlePromise would advance the thread to the next block
itself. Now, it sets the thread's "resolved" flag. This means that when
the promise resolves and the thread is set back to STATUS_RUNNING, the
sequencer will call `execute` on the block a second time, where it will
clear the "resolved" flag, report any values, and return without
actually executing the block a second time, causing the sequencer to
advance the thread to the next block instead.

This has the effect of moving all control flow handling into the
sequencer, making it easier to change in the future and reducing the
chance of logic bugs in diverging code paths.

adroitwhiz committed:
refactor: retire threads immediately in stepHat

This adds a check in stepHat, after moving to the next block, whether
the next block exists. If it doesn't, then unwrap the stack. This means
that a STATUS_RUNNING thread will always have a block on the stack,
since we also removed the other case where a thread could be
STATUS_RUNNING with nothing on the stack--promises--in the previous
commit.

This means that stepHat may cause threads to retire one tick sooner,
and we expect hat blocks with an empty stack to run for at least one
tick, so we also change the runtime to only call stepHat on
edge-activated hat blocks, matching 2.0.

adroitwhiz committed:
refactor: remove null stack check from stepThread

Now that we've removed all cases where a STATUS_RUNNING thread has a
null at the top of the stack, we can get rid of the special case and
simplify the logic.

adroitwhiz committed:
refactor: simplify stepToBranch

We always pass in branchNum, so there's no need to default it to 1.
In addition, branchId will never be undefined--always a string or null.
We can thus remove the "if (!branchId)" check.

adroitwhiz committed:
refactor: simplify unwrapStack

It's easier to understand with early returns.

adroitwhiz committed:
refactor: unify thread step logic into stepOnce

Add the stepOnce function, which is used for stepping a thread once.
We now call that function instead of stepHat, since it implements the
same logic but more completely. This lets us remove stepHat, and
unwrapStack is once again only used in one place and can be inlined to
make it more clear what the sequencer is actually doing.

adroitwhiz committed:
refactor: remove YIELD check from stepThreads

We now clear a thread's YIELD status after stepping an edge-activated
hat block. This allows us to maintain the invariant that a thread will
never be STATUS_YIELD outside of stepThread, and we don't have to
check for it in stepThreads.

adroitwhiz committed:
refactor: remove Sequencer.activeThread

This member variable exists so that when Runtime.stopAll is called in
the middle of a thread, that thread will be stopped before executing any
more blocks while all other threads can continue for an extra tick. It's
set to null at the end of stepThreads, so any calls to Runtime.stopAll
not made from the "stop all" block function will see activeThread as
being null. As such, we can change the block utility "stop all" function
to simply stop the thread itself, and remove this special case from the
more generalized Runtime.stopAll. This removes the need to keep track of
the active thread.

adroitwhiz committed:
refactor: inline stepThread, simplify warp timer

We initialize the warp timer whenever we step into a warp-mode procedure
or the sequencer steps into a warp-mode thread that previously yielded.
Instead of making a new timer instance each time, we can just reuse one
and reset it whenever we enter warp mode.

This removes some complexity from stepThread, and it's now small enough
that it makes more sense to just inline it into stepThreads.

adroitwhiz committed:
refactor: remove _stopThread

We have access to the sequencer and can just call retireThread directly,
instead of needing a one-line paper-thin wrapper function.

adroitwhiz committed:
refactor: simplify monitor script handling

Previously, we ping-ponged from Runtime._pushMonitors ->
Blocks.runAllMonitored -> Runtime.addMonitorScript.
Now, this is all done within _pushMonitors, making the logic a lot
more clear and removing indirection. Yes, there are two nested loops--
I see that as a benefit, as it makes it more clear that there are
(preexisting) performance implications.

adroitwhiz committed:
refactor: remove DONE check from _pushMonitors

Threads will not be set to STATUS_DONE outside of _stepThreads,
and _stepThreads will remove all STATUS_DONE threads from
Runtime.threads before any other function can observe them. Because
_pushMonitors is executed outside of _stepThreads, it never has to
check if a thread is DONE.

adroitwhiz committed:
refactor: remove stackFrame null check from isActiveThread

If the thread's status is not STATUS_DONE, it will have a stack frame.
Note that the inverse is not true--some threads may be STATUS_DONE but
still have a stack frame. We thus choose to keep only the STATUS_DONE
check.

adroitwhiz committed:
refactor: remove isActiveThread and Runtime.threads check

By maintaining the invariant that a STATUS_DONE thread will never exit
STATUS_DONE, we can ensure that any thread removed from Runtime.threads
will always be STATUS_DONE.

This allows us to inline the one remaining check from isActiveThread and
remove the method altogether.

adroitwhiz committed:
refactor: move waiting-threads logic to BlockUtility

It was the same code duplicated across two blocks--seems like a good
candidate for abstraction.

adroitwhiz committed:
refactor: immediately go to next block in stopThisScript

Instead of the hack of pushing a null block onto the stack so that the
sequencer treats it like the end of a substack, we just advance to the
next block directly. Since stopThisScript is only called by executing
blocks, the sequencer will retire the thread if the next block is null,
maintaining the invariants we rely upon.

adroitwhiz committed:
refactor: inline handlePromise

This function has shrunk considerably as a result of previous
refactorings, and can be inlined now.

adroitwhiz committed:
refactor: handle broadcast vars in event blocks

execute.js was previously monkey-patching the BROADCAST_INPUT field
to BROADCAST_OPTION. This is because it's the one droppable input
currently in the VM that returns an {id, name} object from the shadow
block, but a scalar value if there's a block plugged into it. execute.js
was massaging this input into an {id, name} object in *all* cases.

It's much cleaner to simply have the event blocks themselves cast the
argument to the correct "shape", and removes a lot of ugly special-case
code from execute.js.

adroitwhiz committed:
refactor: move thread retiring to Thread

The sequencer doesn't have to do anything when it retires a thread; it
just resets a bunch of the thread's state. The thread is perfectly
capable of doing that itself, and this will eventually allow us to stop
passing the sequencer into handleReport.

adroitwhiz committed:
refactor: clarify "find last reported block" code

adroitwhiz committed:
refactor: pass runtime into handleReport

The remaining uses of `sequencer` in handleReport were simply to reach
into the runtime through it. We can just directly pass the runtime
instead.

adroitwhiz committed:
refactor: inline reuseStackForNextBlock

It was only being used in one place and was only two lines. It's much
easier to just move its logic to the one place that needs it.

adroitwhiz committed:
refactor: initialize thread stack to first block

We were manually setting the first block everywhere we constructed a new
Thread. Instead, just pass it in.

adroitwhiz committed:
refactor: handle BlockCached report with virtual dispatch

We were branching a lot based on static properties--for instance,
whether a block is top-level, a hat block, etc. These do not change
after BlockCached is constructed, so we can define a function which
handles the reported value once when the BlockCached is created and
doesn't have to branch. This also lets us get rid of lastOperation.

adroitwhiz committed:
refactor: clarify promise thaw code

The comments here were a bit terse.

adroitwhiz committed:
refactor: more execute.js code style tweaks

These make the code easier to reason about.

adroitwhiz committed:
refactor: remove allScripts(ByOpcode)Do

There are some nested loops here, but it's better to have them both
in front of you than to wonder what "all scripts" implies.

adroitwhiz committed:
refactor: thread pause/resume

There are now three parts:
- Thread.pause sets the status to PROMISE_WAIT and stores all the data
  needed to resume execution.
- Thread.resume is called by promises that resolve, and sets the status
  back to PROMISE_RUNNING, sets the "just resolved" flag to true,
  and sets the resolved value to the one returned by the promise.
- Thread.finishResuming is called by execute after it handles the
  reported promise value, and resets all the thread's promise-related
  state.

adroitwhiz committed:
refactor: remove RUNNING check from promise handler

If a block returns a promise, it's not allowed to do anything to the
control flow itself. We can thus assume the thread is still
STATUS_RUNNING.

adroitwhiz committed:
refactor: start branches based on block IDs

Treat branch/C-block inputs as regular named inputs whose values are the
IDs of the first blocks within them. We still need BRANCH_INPUT_PREFIX
because we have no way of knowing the "shape" of the input in the VM,
so the inputs will, for now, need to be named SUBSTACK<n>.

adroitwhiz committed:
refactor: make restartThread more efficient

We know the index of the thread already; why not use it?

adroitwhiz committed:
refactor: track project run status more simply

Instead of calculating the number of monitor threads and subtracting it
from the total number of threads, just check whether any non-monitor
threads exist. We don't need to know how many there are, just whether
any exist. This lets us get rid of _getMonitorThreadCount and convert
the _nonMonitorThreadCount property to a boolean.

adroitwhiz committed:
refactor: make script glows less quadratic

Testing whether an element in a list exists is O(n). Checking whether
that same element exists in a hash set is O(1). By converting
script glows from lists to sets, we can track the differences between
them much faster.

adroitwhiz committed:
refactor: remove stackClick check from _updateGlows

If the script executes, we will request block glow no matter what.
An extra stackClick check is just redundant.

adroitwhiz committed:
refactor: avoid filtering edge-activated hats

Instead of filtering through all hat opcodes, we maintain a separate
list of edge-activated hats.

adroitwhiz committed:
refactor: remove `instance = this` antipattern

This seems to be an ES5 holdover that nobody caught.

adroitwhiz committed:
refactor: allow restarting threads in-place

Eventually, I'd like to be able to restart monitor threads in-place,
avoiding having to re-push them over and over.

adroitwhiz committed:
refactor: improve block glow handling

We now track forceNoGlow in _updateGlows, where we actually care about
it, instead of checking it every time a block executes. This is much
cheaper. In addition, requestScriptGlowInFrame has been removed since
scripts won't glow unless the block glow ID is valid anyway.

adroitwhiz committed:
refactor: set thread blockContainer properly

There was a long-standing check in execute.js that fell back to the
flyout blocks to find a block if it didn't exist in the target's block
container. Instead, let's just properly initialize Thread.blockContainer
to its real value.

adroitwhiz committed:
refactor: remove Thread.peekParentStackFrame

This function is no longer used, and is safe to remove.

adroitwhiz committed:
refactor: use thread.pointer and thread.stackFrame

Thread's internal methods can access pointer and stackFrame directly
instead of going through peekStack and peekStackFrame.

adroitwhiz committed:
refactor: inline _restartThread

Now that we can restart threads in-place, _restartThread was just a thin
wrapper function for that functionality. We can just call Thread.restart
directly and save some hassle.

adroitwhiz committed:
fix: set forceNoGlow: false on flyout blocks

The previous commit causes flyout blocks to no longer glow when clicked,
which is, uh, exactly what we asked for. However, it seems an existing
bug has caused flyout blocks to glow when clicked anyway. To preserve
that behavior, we'll now explicitly ask for it.

towerofnix commented on src/engine/execute.js:
```diff
@@ -360,7 +353,7 @@ class BlockCached {
         // The final operation is this block itself. At the top most block is a
         // command block or a block that is being run as a monitor.
-        if (this._definedBlockFunction) {
+        if (typeof this._blockFunction !== 'undefined') {
```
Note that although the commentary describes using a local variable, here that would-be variable's value is just inlined. That means moving it down from being defined *ages* earlier, and instead in-place, immediately where it's used!

The `this._definedBlockFunction = ...` was out of place to begin with - the nearby comment describes `_isHat` and `_blockFunction`, but not `_definedBlockFunction` (so `_definedBlockFunction` feels tacked on), and the point of the earlier calls was to avoid a dynamic lookup. Since `this._blockFunction` is, by definition, a static value, it's just fine to bring this `typeof _blockFunction !== 'undefined'` call further down - it'll have the same result as being evaluated earlier.

towerofnix commented on src/engine/execute.js:
```diff
@@ -215,12 +209,6 @@ class BlockCached {
-         * @type {object}
-         */
-        this._inputs = Object.assign({}, this.inputs);
-
```
I'm noting that `this.inputs` and `this._inputs` are treated as the same for the later code. This seems to be at odds with the original behavior, where `this.inputs` was literally the original block's `inputs` dictionary, and `this._inputs` was a safely mutable copy. Is it really okay to mutate, with `delete` statements, `cached.inputs`?

Well, yes. The follow-up commits reveal:

bcd1530870b69b73358988afb817e44b21a807af: `getNonBranchInputs` (née `getInputs`) *was* a cached function (note how `Blocks` tries to hit `this._cache.inputs[block.id]`)...

192e499807f577a0cec00353af69c5eead0239af: But in fact there's no need to cache it - `BlockCached` is already in charge of that, more appropriately. No caching means it's a fresh object each time `BlockCached` is called and is never accessed outside of `BlockCached`, and so is safe for `BlockCached` to mutate.

b316ab8450aae5e7ff960019534550005bbd43f5: Except surprise! Then we pass the whole `block` in, directly. Now `BlockCached`, rather than `Blocks`, is responsible for calling `getNonBranchInputs`... meaning it's the one making the assumption that this function returns a new (safely mutable) object each time. [And this assumption is laid out bare.](https://github.com/scratchfoundation/scratch-vm/pull/3924/commits/b316ab8450aae5e7ff960019534550005bbd43f5#diff-891cfb1d04f97b904176b7f772916ea88c39fb030ebadb7a60e0b64e46d5aed4R247-R248)

towerofnix commented on src/engine/sequencer.js:
```diff
@@ -77,7 +77,7 @@ class Sequencer {
         this.timer.start();
         // Count of active threads.
-        let numActiveThreads = Infinity;
+        let activeThreads = true;
```
`activeThreads` is *initially* set to `true`, but only so that the `while` loop below runs at all.

towerofnix commented on src/engine/sequencer.js:
```diff
@@ -96,48 +96,46 @@ class Sequencer {
             }
 
-            numActiveThreads = 0;
+            activeThreads = false;
```
`activeThreads` is set to false here, at the start of each iteration of the step loop. This is also the *only* place where `activeThreads` is set to false. So by definition, it'll only come *out* of the current iteration as `false` if nothing sets it to `true` (before the iteration is completed).

towerofnix commented on src/engine/sequencer.js:
```diff
@@ -96,48 +96,46 @@ class Sequencer {
-                // sure we remove it before the next iteration of all threads.
-                if (activeThread.stack.length === 0 ||
-                    activeThread.status === Thread.STATUS_DONE) {
+                    activeThreads = true;
```
And here's the only place where `activeThreads` is set to true. Keep in mind this is inside a nested `for` loop which iterates over all the threads that were active as of the last iteration of the step loop.

At a high level this means the call to `stepThread(activeThread)` is responsible for setting the current thread's `status` to `Thread.STATUS_DONE`, and all `activeThreads` does is reflect if that result was *not* `STATUS_DONE` for any of the threads, during this iteration of the main `stepThreads` loop.

towerofnix commented on src/engine/sequencer.js:
```diff
@@ -155,12 +153,11 @@ class Sequencer {
-                        doneThreads.push(thread);
                     }
                 }
                 this.runtime.threads.length = nextActiveThread;
```
Now here's the interesting part. This is an "old fashioned" filter which overwrites the array in-place as it iterates over it (it'll always be done with the current item before that item gets overwritten), setting `.length` at the end to "chop off" the garbage items. It's semantically the same as:

```js
doneThreads.push(
  ...this.runtime.threads.filter(({status}) => status === Thread.STATUS_DONE));

this.runtime.threads =
     this.runtime.threads.filter(({status}) => status !== Thread.STATUS_DONE);
```

(Try running through the loop in your head if you need to convince yourself of this!)

So let's rewind to `activeThreads`. Remember that it's set to true if any thread's status *isn't* `STATUS_DONE`.

Now, here's the `while` loop at the very top level of `stepThreads`:

```js
while (this.runtime.threads.length > 0 &&
       activeThreads &&
       /* extra conditions */) {...}
```

What I've been wanting to figure out in the review of this commit is, is it necessary to check `this.runtime.threads.length > 0` here *at all?* Or does `activeThreads` already contain that information for us? Assuming that this loop always iterates at least once (for the moment).

Well the `this.runtime.threads.length = nextActiveThread` here kind of gives it away. It and `activeThreads` sure *sound* synonymous. Is that so? Yes — `threads.length` will be non-zero iff any thread is not `STATUS_DONE`; just the same for `activeThreads`.

Note that this filter runs *after* we've iterated over all active threads - it's the last step in the top-level `stepThreads` loop. If it were somehow possible for a thread's status to be changed to `STATUS_DONE`, *after* it was the one responsible for `activeThreads` becoming true, then there would be trouble — this filter could resolve to zero (no active threads) while `activeThreads` remains true (because nothing sets it to false!). But there just isn't any opportunity for a thread's status to become `STATUS_DONE` — the only thing setting its status to `STATUS_DONE` is inside `stepThread(activeThread)`.

So yes, `this.runtime.threads.length > 0` really does become synonymous with `activeThreads` after the first iteration.

<hr>

Now for the takeaway! To simplify this logic, how about we only check `this.runtime.threads.length > 0` on the first iteration? A simple way to do this would be to just change the initial definition for `activeThreads`:

```diff
-let activeThreads = true;
+let activeThreads = this.runtime.threads.length > 0;
```

Et voila: the `while` loop will still neatly not perform its first iteration, because `activeThreads` was appropriately false, and we skip ahead.

But what exactly are we skipping *to?* Well, we set `this.activeThread = null` and return `doneThreads` below - that's everything. If it's possible, it would be cleaner to simply use an early return. Here's the naive example:

```js
if (this.runtime.threads.length === 0) {
    this.activeThread = null;
    return doneThreads;
}
```

That would work, but we'd like to determine the results of those two steps if the `while` loop was outright skipped - if we do, we might be able to simplify this early return, or move it higher up in the function (instead of right before the main `while` loop).

So for the first line: `this.activeThread = null` is purely a clean-up line. In other words it guarantees that `activeThread` is null once `stepThreads` exits, and so only has a value for the duration of `stepThreads`.

But `stepThreads` is also the only code *setting* `activeThread`... so we know that if it *was* null *before* any code evaluated, it doesn't need to be reset to null during the early return either.

```js
if (this.runtime.threads.length === 0) {
    return doneThreads;
}
```

*(Note: `activeThread` was outright made into a local variable later, with the mentioned cleanup removed. Of course, cleanup remains unneeded in the early return.)*

What's that `doneThreads` value going to be? Not much surprise here: it's just going to be an empty array! The only code which pushes to it is inside the main `while` loop, so if the `while` loop is skipped, nothing gets put in it.

```js
if (this.runtime.threads.length === 0) {
    return [];
}
```

This can be moved *all the way to the top* of `stepThreads`. (Or past the side effects `this.timer.start()` and `this.runtime.updateCurrentMSecs()`, if those have any impact outside of `stepThreads`. And only if!)

An early return for the "no active threads" case means we *know* we're going to run the thread loop at least once. So the final clean-up I'd do (while keeping the general structure of `stepThreads` the same) is to flat-out replace the `while` loop with a `do`-`while`:

```js
let activeThreads;
do {
    activeThreads = false;
    /* all the usual logic */
} while (activeThreads &&
         this.timer.timeElapsed() < WORK_TIME &&
         (this.runtime.turboMode || !this.runtime.redrawRequested))
```

Yes, it's true — `do`-`while` loops *are* spooky scary skeletons. But it really does make sense to move this conditional to the end, even beyond the semantic fact that it's "fine" to skip it the very first time — see, the only case where the conditional could *possibly* be true is *after* the some an iteration is complete! That's clear as much when you actually read the conditional parts:

* There only might *not* be `activeThreads` after we've evaluated all the were-active threads and they turned out to become done.
* The work timer could only possibly have elapsed if, you know, we've actually done any work.
* A redraw could only have been requested if some block requested it... as a result of... stepping a thread!

I haven't read ahead so who knows how much `stepThreads` is going to be restructured further. But those are my suggestions based on its current state, and thinking of how I'd refactor if I were the one doing the PR!

towerofnix reviewed (commented):
This is truly fantastic work. I've left a few comments as I'm going through this - since I left a longer commentary on `Sequencer.stepThreads` suggesting further refactoring, I figured I'd share that now before continuing through more.

FurryR commented on src/engine/execute.js:
```diff
@@ -500,65 +399,49 @@ const execute = function (sequencer, thread) {
-                }
+            // Resume thread after the promise resolves
+            const generation = thread.generation;
+            primitiveReportedValue
+                .catch(rejectionReason => {
+                    // Promise rejected: the primitive had some error.
+                    // Log it and proceed.
+                    log.warn('Primitive rejected promise: ', rejectionReason);
+                    // Return an empty string
+                    return '';
+                }).then(resolvedValue => {
```
`catch` on PromiseLike seems wrong.

p.s.: Turbowarp also had this issue before.

towerofnix replied:
Nothing new in this pull request, it's code that exists in the existing repo: https://github.com/scratchfoundation/scratch-vm/blob/16630ee9dd11b8b939c1a5db1c7ee991cb21b058/src/engine/execute.js#L138-L144

Almost certainly out of scope for this pull request to address.

FurryR replied:
It's not.

PromiseLike (`isPromise`) only provides you `then` with `onfulfilled` and `onrejected` parameter, if you are using `catch` method, you are assuming that the PromiseLike will only be `Promise` without assertions.

Original code didn't use `catch`.

towerofnix replied:
Oh, thank you for the correction, that's a good point.

Note [`isPromise`](https://github.com/scratchfoundation/scratch-vm/blob/16630ee9dd11b8b939c1a5db1c7ee991cb21b058/src/engine/execute.js#L31-L37) is defined in this file, but it probably shouldn't be altered, because it is the definition for how block modules are expected to provide promises. Changing `isPromise` could break compatibility with existing extensions (whether part of official scratch-vm or not) and it probably isn't worth it.

adroitwhiz replied:
Fixed this by replacing it with the 2-argument `then` syntax.

adroitwhiz committed:
fix: use Promise#then rejection handler

The isPromise check only ensures that the given item is a "thenable"
(aka that it has a "then" method). This does not guarantee that it's a
real Promise with a "catch" method, and in fact some Scratch addons and
extensions may return fake Promises that don't implement "catch".
